In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib notebook

# Wranglin' – Corralling Unruly Data
One bit at a time
=====

**Abridged Version 0.2.5**

By B Scott, Northwestern/CIERA
adapted from Version by A Miller
May 2024

For this exercise you will need a five different text files. They have been compiled into a tarball that you should [download](https://arch.library.northwestern.edu/downloads/8g84mm66j?locale=en) and unpack in the same directory as this notebook. 

If data are constants, and constants don't change, then we should probably be sure that our data storage solutions do not alter the data in any way.

Within the data science community, the python pandas package is particularly popular for reading, writing, and manipulating data (there will be several examples this week).

The pandas docs state the read_csv() method is the workhorse function for reading text files. So, does pandas "maintain the constant nature of data"?

**Problem 1a**

Create a numpy array, called nums, of length 10000 filled with random numbers. Create a pandas Series object, called s, based on that array, and then write the Series to a file called tmp.txt using the to_csv() method.

Hint - you'll need to name the Series and add the header=True option to the to_csv() call.

In [4]:
nums = np.random.random(1000)
s = pd.DataFrame(nums)
s.to_csv('tmp.txt')

**Problem 1b**

Using the pandas read_csv() method, read in the data to a new variable, called s_read. Do you expect s_read and nums to be the same? Check whether or not your expectations are correct.

Hint - take the sum of the difference not equal to zero to identify if any elements are not the same.

In [5]:
s_read = pd.read_csv('tmp.txt')

In [13]:
np.sum(s_read.values[:,1]-nums)

-1.6353129787816112e-14

So, it turns out that ∼23% of the time, pandas does not in fact read in the same number that it wrote to disk.

The truth is that these differences are quite small (see next slide), but there are many mathematical operations (e.g., subtraction of very similar numbers) that may lead these tiny differences to compound over time such that your data are not, in fact, constant.

In [14]:
print(np.max(np.abs(nums - s_read.values[:,1])))

1.1102230246251565e-16


So, what is going on?

Sometimes, when you convert a number to ASCII (i.e. text) format, there is some precision that is lost in that conversion.

How do you avoid this?

One way is to directly write your files in binary. To do so has serveral advantages: it is possible to reproduce byte level accuracy, and, binary storage is almost always more efficient than text storage (the same number can be written in binary with less space than in ascii).

The downside is that developing your own procedure to write data in binary is a pain, and it places strong constraints on where and how you can interact with the data once it has been written to disk.

Fortuantely, we live in a world with pandas. All this hard work has been done for you, as pandas naturally interfaces with the hdf5 binary table format. (You may want to also take a look at pyTables)

(Historically astronomers have used FITS files as a binary storage solution)

**Problem 1c**

Repeat your procedure from above, but instead of writing to a csv file, use the pandas to_hdf() and read_df() method to see if there are any differences in s and s_read.

Hint - You will need to specify a name for the table that you have written to the hdf5 file in the call to to_hdf() as a required argument. Any string will do.

Hint 2 - Use s_read.values instead of s_read['nums'].values.

In [22]:
s.to_hdf('tmp.txt', key='nums', mode='w')

In [23]:
s_read = pd.read_hdf('tmp.txt')

In [24]:
np.sum(s_read.values-nums)

3.439026841078885e-12

So, if you are using pandas anyway, and if you aren't using pandas –– check it out!, then I strongly suggest removing csv files from your workflow to instead focus on binary hdf5 files. This requires typing the same number of characters, but it ensures byte level reproducibility.

And reproducibiliy is the pillar upon which the scientific method is built.

Is that the end of the story? ... No.

In the previous example, I was being a little tricky in order to make a point. It is in fact possible to create reproducible csv files with pandas. By default, pandas sacrifices a little bit of precision in order to gain a lot more speed. If you want to ensure reproducibility then you can specify that the float_precision should be round_trip, meaning you get the same thing back after reading from a file that you wrote to disk.

In [28]:
s.to_csv('tmp.txt', header=True, index=False)

s_read = pd.read_csv('tmp.txt', float_precision='round_trip')

sum(nums - s_read.values[:,0] != 0)

0

So was all of this in service of a lie?

No. What I said before remains true - text files do not guarantee byte level precision, and they take more space on disk. Text files have some advantages:

anyone, anywhere, on any platform can easily manipulate text files
text files can be easily inspected (and corrected) if necessary
special packages are needed to read/write in binary
binary files, which are not easily interpretable, are difficult to use in version control (and banned by some version control platforms)
To summarize, here is my advice: think of binary as your (new?) default for storing data.

But, as with all things, consider your audience: if you are sharing/working with people that won't be able to deal with binary data, or, you have an incredibly small amount of data, csv (or other text files) should be fine.

## Problem 2) The Sins of Our Predecessors

If at any point in your career you need to access archival infrared data, you will likely need to retrieve that information from the [NASA IPAC InfraRed Science Archive](https://irsa.ipac.caltech.edu). IRSA houses the data for every major NASA IR mission, and several ground-based missions as well (e.g., 2MASS, IRTF). Whether you are sudying brown dwarfs, explosive transients, solar system objects, star-formation, galaxy evolution, Milky Way dust and the resulting extinction of extragalactic observations, or quasars (and much more) the IR plays a critical role.

Given the importance of IR observations, it stands to reason that IRSA would provide data in a simple to read format for modern machines, such as comma separated values or FITS binary tables...

Right?...

**Right?...**

In fact, IRSA has created their own standard for storing data in a text file. The particulars of this format can be found in `irsa_catalog_WISE_iPTF14jg_search_results.tbl`, a file that is written in the standard IRSA format.

*shameless plug for Adam alert!* iPTF14jg is a [really strange star](https://arxiv.org/pdf/1901.10693.pdf) that exhibited a large outburst that we still don't totally understand. The associated data file includes [NEOWISE](https://neowise.ipac.caltech.edu/) observations of the mid-IR evolution of this outburst.

**Problem 2a**

Using `pandas` read the data in the IRSA table file into a `DataFrame` object.

*Hint 1* - you absolutely should look at the text file to develop a strategy to accomplish this goal.

*Hint 2* - you may want to manipulate the text file so that it can more easily be read by `pandas`. **If you do this** be sure to copy the file to another name as you will want to leave the original intact. 

In [35]:
# Solution 1 - pure python solution with pandas

with open('dsfp_wrangling/irsa_catalog_WISE_iPTF14jg_search_results.tbl') as f:
    ll = f.readlines()
    for linenum, l in enumerate(ll):
        if l[0] == '|':
            header = l.replace('|', ',').replace(' ', '')
            header = list(header[1:-2].split(','))
            break
            
irsa_tbl = pd.read_csv("dsfp_wrangling/irsa_catalog_WISE_iPTF14jg_search_results.tbl", 
            skiprows=linenum+4, delim_whitespace=True, 
            header=None, names=header)
irsa_tbl.head(5)

ra        dec   sigra  sigdec  sigradec  w1mpro  w1sigmpro  w1snr  \
0  40.125566  60.879306  0.0947  0.0936   -0.0337  13.251      0.038   28.8   
1  40.125574  60.879297  0.0925  0.0874   -0.0397  13.241      0.036   30.0   
2  40.125623  60.879318  0.0808  0.0794    0.0098  12.753      0.039   27.8   
3  40.125553  60.879286  0.0663  0.0608   -0.0156  12.774      0.028   38.6   
4  40.125572  60.879280  0.0645  0.0599   -0.0223  12.840      0.028   38.7   

   w1rchi2  w2mpro  ...  w4sigmpro_allwise  tmass_key  j_m_2mass  \
0   0.8767  12.735  ...                NaN        NaN        NaN   
1   1.2610  12.629  ...                NaN        NaN        NaN   
2   0.7640  11.901  ...                NaN        NaN        NaN   
3   1.3610  11.935  ...                NaN        NaN        NaN   
4   1.5550  11.897  ...                NaN        NaN        NaN   

   j_msig_2mass  h_m_2mass h_msig_2mass k_m_2mass  k_msig_2mass      dist  \
0           NaN        NaN          NaN       NaN           NaN  0.028668   
1           NaN        NaN          NaN       NaN           NaN  0.048011   
2           NaN        NaN          NaN       NaN           NaN  0.079498   
3           NaN        NaN          NaN       NaN           NaN  0.097389   
4           NaN        NaN          NaN       NaN           NaN  0.107707   

        angle  
0  242.330013  
1  192.659014  
2   69.867298  
3  209.287467  
4  187.006050  

[5 rows x 37 columns]

That pure python solution is a bit annoying as it requires a for loop with a break, and specific knowledge about how IRSA tables handle data headers (hence the use of `linenum + 4` for `skiprows`). Alternatively, one could  manipulate the data file to read in the data.

In [ ]:
# solution 2 - edit the text file
# !cp irsa_catalog_WISE_iPTF14jg_search_results.tbl tmp.tbl
### delete lines 1-89, and 90-92
### replace whitespace with commas (may require multiple operations)
### replace '|' with commas
### replace ',,' with single commas
### replace ',\n,' with '\n'
### delete the comma at the very beginning and very end of the file

tedit_tbl = pd.read_csv('tmp.tbl')
tedit_tbl.head(5)

That truly wasn't all that better - as it required a bunch of clicks/text editor edits. (There are programs such as `sed` and `awk` that could be used to execute all the necessary edits from the command line, but that too is cumbersome and somewhat like the initial all `python` solution). 

If astronomers are creating data in a "standard" format, then it ought to be easy for other astronomers to access that data.

Fortunately, in this particular case, there is an easy solution - [`astropy Tables`](http://docs.astropy.org/en/stable/table/). 

IRSA tables are so commonly used throughout the community, that the folks at `astropy` have created a convenience method for all of us to read in tables created in that particular (unusual?) format.

**Problem 2b**

Use [`Table.read()`](http://docs.astropy.org/en/stable/api/astropy.table.Table.html#astropy.table.Table.read) to read in `irsa_catalog_WISE_iPTF14jg_search_results.tbl` to an `astropy Table` object.

In [ ]:
from astropy.table import Table

Table.read('irsa_catalog_WISE_iPTF14jg_search_results.tbl', format='ipac')

A benefit to using this method, as opposed to `pandas`, is that data typing and data units are naturally read from the IRSA table and included with the associated columns. Thus, if you are uncertain if some brightness measurement is in magnitudes or Janskys, the `astropy Table` can report on that information.

Unfortunately, `astropy` does *not* know about every strange formating decision that every astronomer has made at some point in their lives (as we are about to see...) 

## Problem 3) The Sins of Our Journals

Unlike IRSA/IPAC, which uses a weird but nevertheless consistent format for data tables, data retrieved from Journal articles essentially follows no rules. In principle, tables in Journal articles are supposed to be provided in a machine readable format. In practice, as we are about to see, this is far from the case.

For this particular wrangling case study we will focus on supernova light curves, a simple thing to report: time, filter, brightness, uncertainty on that brightness, that the community has nevertheless managed to mangle into some truly wild and difficult to parse forms.

(Sorry for the heavy emphasis on time-domain examples - I'm pulling straight from my own life today, but the issues described here are not perfectly addressed by any subfield within the astro umbrella)

Here is the LaTeX-formatted version of Table 4 from [Miller et al. 2011](https://iopscience.iop.org/article/10.1088/0004-637X/730/2/80/meta):

<img style="display: block; margin-left: auto; margin-right: auto" src="images/Miller11_tbl4.png" width="350" align="middle">

That is a very simple table to interpret, no?

Have a look at the ["machine-readible" file](https://iopscience.iop.org/0004-637X/730/2/80/suppdata/apj382770t4_ascii.txt?doi=10.1088/0004-637X/730/2/80) that ApJ provides for readers that might want to evaluate these photometric measurements.

**Problem 3a** 

Read the ApJ version of Table 4 from from Miller et al. 2011 – called `Miller_et_al2011_table4.txt` – into either a `pandas DataFrame` or an `astropy Table`.

In [40]:
# pure python solution with pandas

tbl4 = pd.read_csv('dsfp_wrangling/Miller_et_al2011_table4.txt', 
                   skiprows=4, delim_whitespace='	',
                   skipfooter=2, engine='python',
                #    names=['t_mid^a','J mag','H mag','K_s mag']
                   )
# tbl4.drop(columns=[], inplace=True)

# print(tbl4)

In [46]:
tbl4['(MJD)']

55466.137  10.04  +or-  0.03  9.14  +or-    0.03
55468.145  9.99   +or-  0.03  9.06  +or-    0.04
55469.148  10.04  +or-  0.03  9.07  +or-    0.03
55479.109  10.11  +or-  0.03  9.11  +or-    0.03
55504.164  10.20  +or-  0.03  9.24  +or-    0.03
55513.195  10.29  +or-  0.03  9.34  +or-    0.03
55518.168  10.32  +or-  0.03  9.34  +or-    0.04
55527.117  10.35  +or-  0.03  9.40  +or-    0.03
55531.145  10.40  +or-  0.03  9.44  +or-    0.03
55543.066  10.45  +or-  0.03  9.48  +or-    0.04
Name: (MJD), dtype: float64

In [43]:
tbl4.drop(tbl4.columns[2,5,8], inplace=True, axis=1)

IndexError: too many indices for array: array is 1-dimensional, but 3 were indexed

That wasn't too terrible. But what if we consider a more typical light curve table, where there are loads of missing data, such as Table 2 from [Foley et al. 2009](https://iopscience.iop.org/article/10.1088/0004-6256/138/2/376#aj309430t2):

Again, this table is straightforward to read, and it isn't hard to imagine how one could construct a machine-readable csv or other file from this information. But alas, this is not what is available from ApJ. So, we will need to figure out how to deal with both the missing data, "...", and the weird convention that many astronomers use where the uncertainties are (a) not reported in their own column, and (b) are not provided in the same units as the measurement itself. I can understand the former, but the later is somewhat baffling...

**Problem 3b** 

Read the ApJ version of Table 2 from from Foley et al. 2009 – called `Foley_et_al2009_table2.txt` – into either a `pandas DataFrame` or an `astropy Table`.

In [ ]:
# a (not terribly satisfying) pure python solution
# read the file in, parse and write another file that plays nice with pandas

with open('Foley_et_al2009_for_pd.csv','w') as fw:
    print('JD,Bmag,Bmag_unc,Vmag,Vmag_unc,Rmag,Rmag_unc,Imag,Imag_unc,Unfiltmag,Unfiltmag_unc,Telescope',file=fw)
    with open('Foley_et_al2009_table2.txt') as f:
        ll = f.readlines()
        for l in ll:
            if l[0] == '2':
                print_str = l.split()[0] + ','
                for col in l.split()[1:]:
                    if col == 'sdotsdotsdot':
                        print_str += 
                    #complete 
                    else:
                        print_str += '{},'.format(col)

                print(print_str,file=fw)

pd.read_csv('Foley_et_al2009_for_pd.csv')

Okay - there is nothing elegant about that particular solution. But it works, and wranglin' ain't pretty. 

It is likely that you developed a solution that looks very different from this one, and that is fine. When data are provided in an unrulely format, the most important thing is to develop some method, any method, for converting the information into a useful format. Following whatever path you used above, it should now be easy to plot the light curve of SN 2008ha.

## Problem 4) My Heart Will Go On

Sometimes there is no difficultly whatsoever in reading in the data (as was the case in **Problems 2** and **3**), but instead the difficultly lies in wranglin' the data to be appropriate for the model that you are building.

For the next problem we will work with the famous [Titanic survival](https://www.kaggle.com/c/titanic/data?) data set.

Briefly, [the Titantic](https://thefilmcricket.files.wordpress.com/2012/04/film-titanic_clar.jpg) is a [famous](https://wallpapercave.com/wp/jrF8rQK.jpg) historical ship that was thought to be unsinkable. **Spoiler alert** it hit an iceberg and sank. The data in the Titanic data set includes information about 891 passengers from the Titanic, as well as whether or not they survived. The aim of this data set is to build a machine learning model to predict which passengers survived and which did not.

The features include: 

|Feature    | Description |
|:---------:|:--------------------------------------:|
|PassengerId| Running index that describes the individual passengers|
|Pclass| A proxy for socio-economic status (1 = Upper class, 2 = Middle Class, 3 = Lower Class)|
|Name| The passenger's name|
|Sex | The passenger's sex|
|Age | The passenger's age - note age's ending in 0.5 are estimated |
|SibSp| The sum of the passenger's sibblings and spouces on board|
|Parch| The sum of the passenger's parents and children on board|
|Ticket| The ticket number for the passenger|
|Fare| The price paid for the ticket by th passenger|
|Cabin| The Cabin in which the passenger stayed|
|Embarked| The point of Origin for the Passenger: C = Cherbourg, S = Southampton, Q = Queenstown|

And of course, we are trying to predict:

|Label    | Description |
|:---------:|:--------------------------------------:|
|Survived| 1 = yes; 0 = no|


**Problem 4a**

Read in the Titanic training data and create the `scikit-learn` standard `X` and `y` arrays to hold the features and the labels, respectively.

In [65]:
titanic_df = pd.read_csv('dsfp_wrangling/titanic_kaggle_training_set.csv')

# X = titanic_df[['PassengerId', 'Fare', 'Cabin']]
# y = titanic_df[['Survived']]

Now that we have the data in the appropriate `X` and `y` arrays, estimate the accuracy with which a [K nearest neighbors](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html) classification model can predict whether or not a passenger would survive the Titanic disaster. Use $k=10$ fold cross validation for the prediction.

In [66]:
titanic_df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [67]:
titanic_df = titanic_df.drop(['Name','Ticket','Cabin', 'PassengerId'], 1)

/var/folders/6b/q0ls5s7570d8s04fbvw5gmxw0000gn/T/ipykernel_43109/2967200523.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  titanic_df = titanic_df.drop(['Name','Ticket','Cabin', 'PassengerId'], 1)


In [68]:
titanic_df.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

In [69]:
titanic_df['Embarked'] = titanic_df['Embarked'].fillna('S')
titanic_df['Age'] = titanic_df['Age'].fillna(titanic_df['Age'].mean())

In [70]:
titanic_df

Survived  Pclass     Sex        Age  SibSp  Parch     Fare Embarked
0           0       3    male  22.000000      1      0   7.2500        S
1           1       1  female  38.000000      1      0  71.2833        C
2           1       3  female  26.000000      0      0   7.9250        S
3           1       1  female  35.000000      1      0  53.1000        S
4           0       3    male  35.000000      0      0   8.0500        S
..        ...     ...     ...        ...    ...    ...      ...      ...
886         0       2    male  27.000000      0      0  13.0000        S
887         1       1  female  19.000000      0      0  30.0000        S
888         0       3  female  29.699118      1      2  23.4500        S
889         1       1    male  26.000000      0      0  30.0000        C
890         0       3    male  32.000000      0      0   7.7500        Q

[891 rows x 8 columns]

In [71]:
titanic_df.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

In [72]:
X = titanic_df.drop(['Survived'],1)
y = titanic_df['Survived']

/var/folders/6b/q0ls5s7570d8s04fbvw5gmxw0000gn/T/ipykernel_43109/2661880598.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X = titanic_df.drop(['Survived'],1)


**Problem 4b**

Train a $k=7$ nearest neighbors machine learning model on the Titanic training set.

In [73]:
from sklearn.neighbors import KNeighborsClassifier
knn_clf = KNeighborsClassifier(n_neighbors=7)
knn_clf.fit(X, y)

ValueError: could not convert string to float: 'male'

Note - that should have failed! And for good reason - recall that `kNN` models measure the Euclidean distance between all points within the feature space. So, when considering the sex of a passenger, what is the *numerical* distance between male and female? 

In other words, we need to wrangle this data before we can run the machine learning model. 

Most of the features in this problem are non-numeric (i.e. we are dealing with categorical features), and therefore we need to figure out how to include them in the `kNN` model. 

The first step when wrangling for machine learning is to figure out if anything can be thrown away. We certainly want to avoid including any uninformative features in the model. 

*If you haven't already, now would be a good time to create a new cell and examine the contents of the csv*

In [ ]:
titanic_df

In [74]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
transformed_data = X.select_dtypes(include='object').apply(lambda i: le.fit_transform(i))

In [75]:
X = X.drop(X.select_dtypes(include='object'),axis=1)

In [76]:
X

Pclass        Age  SibSp  Parch     Fare
0         3  22.000000      1      0   7.2500
1         1  38.000000      1      0  71.2833
2         3  26.000000      0      0   7.9250
3         1  35.000000      1      0  53.1000
4         3  35.000000      0      0   8.0500
..      ...        ...    ...    ...      ...
886       2  27.000000      0      0  13.0000
887       1  19.000000      0      0  30.0000
888       3  29.699118      1      2  23.4500
889       1  26.000000      0      0  30.0000
890       3  32.000000      0      0   7.7500

[891 rows x 5 columns]

In [77]:
from sklearn.neighbors import KNeighborsClassifier
knn_clf = KNeighborsClassifier(n_neighbors=7)
knn_clf.fit(X, y)

KNeighborsClassifier(n_neighbors=7)

**Problem 4c** 

Are there any features that *obviously* will not help with this classification problem?

*If you answer yes - ignore those features moving forward*

**Solution 4c**

*write your solution here*


Given that we have both categorical and numeric features, let's start with the numerical features and see how well they can predict survival on the Titanic.

One note - for now we are going to exclude `Age`, because as you saw when you examined the data, there are some passengers that do not have any age information. This problem, known as "missing data" is one that we will deal with before the end of this problem.

**Problem 4d**

How accurately can the numeric features, `Pclass`, `SibSp`, `Parch`, and `Fare` predict survival on the Titanic? Use a $k = 7$ Nearest Neighbors model, and estimate the model accuracy using 10-fold cross validation. 

*Hint 1 - you'll want to redefine your features vector `X`*

*Hint 2 - you may find [`cross_val_score`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html) from `scikit-learn` helpful*

In [78]:
from sklearn.model_selection import cross_val_score


cv_results = cross_val_score(knn_clf, X, y, cv=3)

print('The accuracy from numeric features = {:.2f}%'.format(100*np.mean(cv_results)))

The accuracy from numeric features = 66.22%


An accuracy of 68% isn't particularly inspiring. But, there's a lot of important information that we are excluding. As far as the Titanic is concerned, Kate and Leo taught us that [female passengers are far more likely to survive](https://qph.fs.quoracdn.net/main-qimg-93eb36091c7eec872b891fa51dc5722b), while [male passengers are not](http://hoycinema.abc.es/Media/201602/03/titanic-kate-dicaprio--644x362.jpg). So, if we can include gender in the model then we may be able to achieve more accurate predictions. 

**Problem 4e**

Create a new feature called `gender` that equals 1 for male passengers and 2 for female passengers. Add this feature to your dataframe, and include it in a `kNN` model with the other numeric features. Does the inclusion of this feature improve the 10-fold CV accuracy?

In [ ]:
gender = # complete
# complete

titanic_df['gender'] = gender

X = # complete

cv_results = cross_val_score( # complete

print('The accuracy when including gender = {:.2f}%'.format(100*np.mean(cv_results)))

A 14% improvement is pretty good! But, we can wrangle even more out of the gender feature. Recall that `kNN` models measure the Euclidean distance between sources, meaning the scale of each feature really matters. Given that the fare ranges from 0 up to 512.3292, the `kNN` model will see this feature as far more important than `gender`, for no other reason than the units that have been adopted. 

If women are far more likely to survive than men, then we want to be sure that gender is weighted at least the same as all the other features, which we can do with a minmax scaler. As a brief reminder - a minmax scaler scales all values of a feature to be between 0 and 1 by subtracting the minimum value of each feature and then dividing by the maximum minus the minimum. 

**Problem 4f**

Scale all the features from the previous problem using a minmax scaler and evaluate the CV accuracy of the `kNN` model.

*Hint - you may find [`MinMaxScaler`](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html) helpful*

In [ ]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
scaler.fit(X)
Xminmax = scaler.transform(X)

knn_clf = # complete

cv_results = cross_val_score( # complete

print('The accuracy when scaling features = {:.2f}%'.format(100*np.mean(cv_results)))

Scaling the features leads to further improvement!

Now turn your attention to another categorical feature, `Embarked`, which includes the point of origin for each passenger and has three categories, `S`, `Q`, and `C`. We need to convert these values to a numeric representation for inclusion in the model.

**Problem 4g**

Convert the categorical feature `Embarked` to a numeric representation, and add it to the `titanic_df`.

In [ ]:
# following previous example, set C = 0, S = 1, Q = 2

porigin = # complete
# complete
# complete
# complete

titanic_df['porigin'] = porigin

But wait! Does this actually make sense?

Our "numerification" has now introduced order where there previously was none. We are effectively telling the model that Cherbourg and Queenstown are far apart (not in distance but in terms of the similarity of the passengers that boarded the ship in each location), while each are equally close to Southampton. Is there actually any evidence to support this conclusion? 

By definition categorical features do not have order (e.g., cat, dog, horse, elephant), and therefore we should not impose any when converting these features to numeric values for inclusion in our model. Instead, we should be creating a new set of binary features for every category within the feature set. Thus, `Embarked` will now need to be represented by 3 different features, where the feature `Queenstown` equals one for passengers that boarded there and zero for everyone else. 

**Problem 4h**

Complete the function below that will automatically create binary arrays for a categorical feature.

In [ ]:
def create_bin_cat_feats(feature_array):
    categories = np.unique(feature_array)
    feat_dict = {}
    for cat in categories:
        #complete
    
    return feat_dict

**Problem 4i**

Use the `create_bin_cat_feats` function to convert the `Embarked` and `Sex` categorical features to a numeric representation (yes we need to do this for `Sex` as well where we otherwise previously introduced order). Add these features to the `titanic_df` data frame.

In [ ]:
gender_dict = 
porigin_dict = 

for feat in gender_dict.keys():
    titanic_df[feat] = gender_dict[feat]
    
for feat in porigin_dict.keys():
    titanic_df[feat] = porigin_dict[feat]

**Problem 4j**

Use the newly created `female`, `male`, `S`, `Q`, and `C` features in combination with the `Pclass`, `SibSp`, `Parch`, and `Fare` features to estimate the classification accuracy of a $k = 7$ nearest neighbors model with 10-fold cross validation.

How does the addition of the point of origin feature affect the final model output?

*Hint - don't forget to scale the features in the model*

In [ ]:
from sklearn.preprocessing import MinMaxScaler

X = # complete

scaler = MinMaxScaler()
scaler.fit(X)
Xminmax = scaler.transform(X)

cv_results = cross_val_score( # complete

print('The accuracy with categorical features = {:.2f}%'.format(100*np.mean(cv_results)))

The last thing we'd like to add to the model is the `Age` feature. Unfortunately, for 177 passengers we do not have a reported value for their age. This is a standard issue when building models known as "missing data" and this happens in astronomy all the time (for example, LSST is going to observe millions of L and T dwarfs that are easily detected in the $y$-band, but which will not have a detection in $u$-band).

There are several different strategies for dealing with missing data. The first and most straightforward is to simply remove observations with missing data (note - to simplify this example I already did this by removing the 2 passengers from the training set that did not have an entry for `Embarked`). 

This strategy is perfectly fine if only a few sources have missing information (2/891 for `Embarked` - and none of the test set sources are missing `Embarked`). If, however, a significant fraction are missing data, this strategy would remove a lot of useful data from the model.

If you cannot remove the sources with missing data, then it is essential to ask the following question: 

Does the missing information have meaning? 

In the LSST L/T dwarf example, the lack of a $u$-band detection is meaningful: these stars are too faint for LSST. When this is the case, an indicator value (e.g., -999) allows the model to recognize the non-detection. 

For the Titanic data, the lack of age information is not meaningful. Simply put, there are some passengers that did not have recorded ages. We will now show this to be the case. 

**Problem 4k**

Replace the unknown ages with a value of -999, and estimate the accuracy of the model via 10-fold cross validation.

In [ ]:
age_impute = # complete
# complete

titanic_df['age_impute'] = age_impute

X = # complete

scaler = MinMaxScaler()
scaler.fit(X)
Xminmax = scaler.transform(X)

cv_results = cross_val_score( # complete

print('The accuracy with -999 for missing ages = {:.2f}%'.format(100*np.mean(cv_results)))

The accuracy of the model hasn't improved by adding the age information (even though we know children were more likely to survive than adults). 

Given that the missing ages don't have meaning, we need to develop alternative strategies for "imputing" the missing data. The most simple approach in this regard is to replace the missing values with the mean value of the feature distribution for sources that do have measurements (use the median if the distribution has significant outliers).  

**Problem 4l**

Replace the unknown ages with the mean age of passengers, and estiamte the accuracy of the model via 10-fold cross validation.

In [ ]:
age_impute = # complete
# complete

titanic_df['age_impute'] = age_impute

X = # complete

scaler = MinMaxScaler()
scaler.fit(X)
Xminmax = scaler.transform(X)

cv_results = cross_val_score( # complete

print('The accuracy with the mean for missing ages = {:.2f}%'.format(100*np.mean(cv_results)))